File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        vector<int> mpp(256, -1);
        int maxLen = 0;
        int left = 0, right = 0;

        while(right<n) {
            if(mpp[s[right]] != -1) {
                left = max(mpp[s[right]]+1, left); //check if the char is in the range left-right
            }

            mpp[s[right]] = right;
            maxLen = max(maxLen, (right - left + 1));
            right++;
        }
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int l = 0, r = 0, result = 0;
        while (l <= r && r < s.length()) {
            while (map.containsKey(s.charAt(r))) {
                map.remove(s.charAt(l));
                l++;
            }
            map.put(s.charAt(r), 1);
            r++;
            result = Math.max(r - l, result);
        }
        return result;
    }
}